File: membrane-js/src/swarm.ts
import { Membrane } from './client';

export class Swarm {
  private membrane: Membrane;

  constructor(membrane: Membrane) {
    this.membrane = membrane;
  }

  async plan(payload: any): Promise<any> {
    const url = `${this.membrane.baseUrl.replace(/\/$/, '')}/swarm/plan`;
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.membrane.apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Membrane swarm.plan failed (${response.status}): ${errorText}`);
    }
    return response.json();
  }

  async map(payload: any): Promise<any> {
    const url = `${this.membrane.baseUrl.replace(/\/$/, '')}/swarm/map`;
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.membrane.apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Membrane swarm.map failed (${response.status}): ${errorText}`);
    }
    return response.json();
  }
}

File: membrane-py/membrane/swarm.py
import requests

class Swarm:
    def __init__(self, membrane_client):
        self.membrane = membrane_client

    def plan(self, payload):
        """
        Call /v1/swarm/plan to forecast cost, latency, and risk before running the swarm.
        """
        url = f"{self.membrane.base_url.rstrip('/')}/swarm/plan"
        headers = {
            "Authorization": f"Bearer {self.membrane.api_key}",
            "Content-Type": "application/json"
        }
        res = requests.post(url, json=payload, headers=headers)
        res.raise_for_status()
        return res.json()

    def map(self, payload):
        """
        Call /v1/swarm/map to run a parallel map-reduce job on documents/logs with strong isolation.
        """
        url = f"{self.membrane.base_url.rstrip('/')}/swarm/map"
        headers = {
            "Authorization": f"Bearer {self.membrane.api_key}",
            "Content-Type": "application/json"
        }
        res = requests.post(url, json=payload, headers=headers)
        res.raise_for_status()
        return res.json()
